pkg/aws/plugin.go

```diff
@@ -0,0 +1,258 @@
+				TagSpecifications: getCreateTagSpecifications(req.Deployment.Namespace, vpcName, types.ResourceTypeVpc),
+			}
+			createVpcOutput, err := ec2Client.CreateVpc(ctx, createVpcInput)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create VPC: %w", err)
+			}
+			vpc = createVpcOutput.Vpc
+
+			// Wait until default security group is available
+			securityGroupExistsWaiter := ec2.NewSecurityGroupExistsWaiter(ec2Client)
+			err = securityGroupExistsWaiter.Wait(ctx, &ec2.DescribeSecurityGroupsInput{
+				Filters: []types.Filter{
+					{Name: aws.String("vpc-id"), Values: []string{*createVpcOutput.Vpc.VpcId}},
+				},
+			}, 10*time.Second)
+			if err != nil {
+				return nil, fmt.Errorf("unable to wait for default security group: %w", err)
+			}
+
+			// Remove default inbound and outbound rules from default security group
+			getSecurityGroupInput := &ec2.DescribeSecurityGroupsInput{
+				Filters: []types.Filter{
+					{Name: aws.String("vpc-id"), Values: []string{*vpc.VpcId}},
+				},
+			}
+			describeSecurityGroupsOutput, err := ec2Client.DescribeSecurityGroups(ctx, getSecurityGroupInput)
+			if err != nil {
+				return nil, fmt.Errorf("unable to get security groups: %w", err)
+			}
+			if len(describeSecurityGroupsOutput.SecurityGroups) == 1 {
+				securityGroup := describeSecurityGroupsOutput.SecurityGroups[0]
+				revokeSecurityGroupIngressInput := &ec2.RevokeSecurityGroupIngressInput{
+					GroupId: securityGroup.GroupId,
+					IpPermissions: []types.IpPermission{
+						{
+							IpProtocol: aws.String("-1"),
+							UserIdGroupPairs: []types.UserIdGroupPair{{
+								GroupId: securityGroup.GroupId,
+								UserId:  aws.String(req.Deployment.Id),
+							}},
+						},
+					},
+				}
+				_, err = ec2Client.RevokeSecurityGroupIngress(ctx, revokeSecurityGroupIngressInput)
+				if err != nil {
+					return nil, fmt.Errorf("unable to revoke default inbound security group rule: %w", err)
+				}
+				revokeSecurityGroupEgressInput := &ec2.RevokeSecurityGroupEgressInput{
+					GroupId: securityGroup.GroupId,
+					IpPermissions: []types.IpPermission{
+						{
+							IpProtocol: aws.String("-1"),
+							IpRanges:   []types.IpRange{{CidrIp: aws.String("0.0.0.0/0")}},
+						},
+					},
+				}
+				_, err = ec2Client.RevokeSecurityGroupEgress(ctx, revokeSecurityGroupEgressInput)
+				if err != nil {
+					return nil, fmt.Errorf("unable to revoke default outbound security group rule: %w", err)
+				}
+			}
+		}
+
+		// Get existing subnet
+		subnetName := getSubnetName(req.Deployment.Namespace, region)
+		describeSubnetsOutput, err := ec2Client.DescribeSubnets(ctx, &ec2.DescribeSubnetsInput{
+			Filters: getDescribeFilter(req.Deployment.Namespace, subnetName),
+		})
+		if err != nil {
+			return nil, fmt.Errorf("unable to get subnets: %w", err)
+		}
+		if len(describeSubnetsOutput.Subnets) == 1 {
+			subnet = &describeSubnetsOutput.Subnets[0]
+			if subnet.AvailabilityZone != nil && *subnet.AvailabilityZone != availabilityZone {
```

Comment:
Does this mean we need a subnet for each AZ? If so, shouldn't we add that to the filter for getting them or something? / Create a new subnet in each new AZ we create a resource in?

Reply:
Yea this was a little more involved than what I thought was needed for the initial version. I would need to split IP address into the different availability zones on the very first VM creation. This IP address range split cannot be changed after the first VM creation. Since the user needs to specify which availability zone they want to create a VM in, there may be an issue where a user wants to put a bulk of their VMs in a particular availability zone but it doesn't have any addresses left because the address range has been evenly split across all the other availability zones.

An argument could be made that user doesn't specify the particular availability zone but rather the region which is broader. I would also like this approach because then we can do the load balancing across different availability zones. The only question is where would the user specify the region? There is no region parameter in RunInstances ([source](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html)), only availability zone.

Reply:
Okay yeah, given this I think the option that makes the most sense (for now) is to create subnets as needed when VMs specify a new AZ and get a new address space for each. Since the address space allocation now allows you to get variable size prefixes, we can make them whatever size makes sense.

Reply:
Issue at #474